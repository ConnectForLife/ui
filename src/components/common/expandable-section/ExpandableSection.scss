#app .expandable-section {
  padding: 1.5em;
  margin-top: 2em;
  background-color: #fafafa;

  .expandable-section-header {
    display: flex;
    justify-content: space-between;
    align-items: center;

    .icon-buttons {
      display: flex;
      justify-content: flex-end;

      .icon-button {
        padding: 4px;
        font-size: 16px !important;

        &:not(:last-child) {
          margin-right: 1em;
        }
      }
    }
  }

  .inline-fields {
    display: flex;

    .input-container {
      min-width: 120px;

      &:not(:last-of-type) {
        margin-right: 1em;
      }
    }

    .flex-1 {
      flex: 1;
    }
  }
}
