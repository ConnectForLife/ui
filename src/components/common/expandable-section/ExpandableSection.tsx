import React, { useEffect, useRef, useState } from 'react';
import './ExpandableSection.scss';
import { Button } from 'reactstrap';

export default function ExpandableSection({
  headerComponent,
  disabledHeaderComponent = headerComponent,
  bodyComponent,
  isRemovable = false,
  onRemove = null,
  isExpandTriggered = false
}) {
  const [isOpen, setIsOpen] = useState(false);
  const prevIsExpandTriggeredRef = useRef<Boolean>();
  const prevIsOpenRef = useRef<Boolean>();
  useEffect(() => {
    prevIsExpandTriggeredRef.current = isExpandTriggered;
    prevIsOpenRef.current = isOpen;
  });
  const isExpandTriggeredChanged = isExpandTriggered !== prevIsExpandTriggeredRef.current;
  const isOpenChanged = isOpen !== prevIsOpenRef.current;
  if (!isOpenChanged && isExpandTriggeredChanged && isOpen !== isExpandTriggered) {
    setIsOpen(isExpandTriggered);
  }
  const isExpanded = isOpenChanged ? isOpen : isExpandTriggeredChanged ? isExpandTriggered : isOpen;
  return (
    <div className="expandable-section">
      <div className="expandable-section-header">
        <div className="inline-fields col-6 pl-0">{isExpanded ? headerComponent : disabledHeaderComponent}</div>
        <div className="icon-buttons col-6">
          {isRemovable && !!onRemove && (
            <Button className="icon-button" onClick={onRemove} data-testid="removeExpandableSectionButton">
              <i className="bi bi-trash" />
            </Button>
          )}
          <Button className="icon-button" onClick={() => setIsOpen(!isExpanded)} data-testid="expandableSectionButton">
            <i className={`bi ${isExpanded ? 'bi-arrows-collapse' : 'bi-arrows-expand'}`} />
          </Button>
        </div>
      </div>
      {isExpanded && <div className="expandable-section-body">{bodyComponent}</div>}
    </div>
  );
}
