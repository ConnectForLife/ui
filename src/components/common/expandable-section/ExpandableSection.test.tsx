import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import ExpandableSection from './ExpandableSection';
import { Input } from 'reactstrap';

function emptyFunction() {
  return null;
}

test('should render match snapshot', () => {
  const headerComponent = <Input value="test1" onChange={emptyFunction} />;
  const bodyComponent = (
    <div>
      <Input value="test2" onChange={emptyFunction} />
      <Input value="test3" onChange={emptyFunction} />
    </div>
  );
  const { container } = render(<ExpandableSection headerComponent={headerComponent} bodyComponent={bodyComponent} />);
  expect(container).toMatchSnapshot();
});

test('should deletable render match snapshot', () => {
  const headerComponent = <Input value="test1" onChange={emptyFunction} />;
  const bodyComponent = (
    <div>
      <Input value="test2" onChange={emptyFunction} />
      <Input value="test3" onChange={emptyFunction} />
    </div>
  );
  const { container } = render(
    <ExpandableSection headerComponent={headerComponent} bodyComponent={bodyComponent} isRemovable onRemove={emptyFunction} />
  );
  expect(container).toMatchSnapshot();
});

test('should be collapsed and match snapshot', () => {
  const headerComponent = <Input value="test1" onChange={emptyFunction} />;
  const bodyComponent = (
    <div>
      <Input value="test2" onChange={emptyFunction} />
      <Input value="test3" onChange={emptyFunction} />
    </div>
  );
  const { container } = render(<ExpandableSection headerComponent={headerComponent} bodyComponent={bodyComponent} />);
  const header = container.querySelector('.expandable-section-header');
  const body = container.querySelector('.expandable-section-body');
  expect(header).toBeInTheDocument();
  expect(body).not.toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test('should expand and match snapshot', () => {
  const headerComponent = <Input value="test1" onChange={emptyFunction} />;
  const bodyComponent = (
    <div>
      <Input value="test2" onChange={emptyFunction} />
      <Input value="test3" onChange={emptyFunction} />
    </div>
  );
  const { container } = render(<ExpandableSection headerComponent={headerComponent} bodyComponent={bodyComponent} />);
  const expandButton = container.querySelector('.icon-button.btn');
  fireEvent.click(expandButton);
  const header = container.querySelector('.expandable-section-header');
  const body = container.querySelector('.expandable-section-body');
  expect(header).toBeInTheDocument();
  expect(body).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test('should collapse and match snapshot', () => {
  const headerComponent = <Input value="test1" onChange={emptyFunction} />;
  const bodyComponent = (
    <div>
      <Input value="test2" onChange={emptyFunction} />
      <Input value="test3" onChange={emptyFunction} />
    </div>
  );
  const { container } = render(<ExpandableSection headerComponent={headerComponent} bodyComponent={bodyComponent} />);
  const expandButton = container.querySelector('.icon-button.btn');
  fireEvent.click(expandButton);
  fireEvent.click(expandButton);
  const header = container.querySelector('.expandable-section-header');
  const body = container.querySelector('.expandable-section-body');
  expect(header).toBeInTheDocument();
  expect(body).not.toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test('should switch between disabled header component and header component on expand/collapse', () => {
  const disabledHeaderTextContent = 'disabled test1';
  const bodyComponent = (
    <div>
      <Input value="test2" onChange={emptyFunction} />
      <Input value="test3" onChange={emptyFunction} />
    </div>
  );
  const { container } = render(
    <ExpandableSection
      disabledHeaderComponent={<div>{disabledHeaderTextContent}</div>}
      headerComponent={<Input value="test1" onChange={emptyFunction} />}
      bodyComponent={bodyComponent}
    />
  );
  let body = container.querySelector('.expandable-section-body');
  const disabledHeader = container.querySelector('.expandable-section-header');
  expect(disabledHeader).toBeInTheDocument();
  expect(disabledHeader).toHaveTextContent('disabled test1');
  expect(body).not.toBeInTheDocument();
  const expandButton = container.querySelector('.icon-button.btn');
  fireEvent.click(expandButton);
  body = container.querySelector('.expandable-section-body');
  const header = container.querySelector('.expandable-section-header');
  expect(header).toBeInTheDocument();
  expect(body).toBeInTheDocument();
});
