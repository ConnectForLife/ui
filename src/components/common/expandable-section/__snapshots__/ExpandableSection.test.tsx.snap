// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be collapsed and match snapshot 1`] = `
<div>
  <div
    class="expandable-section"
  >
    <div
      class="expandable-section-header"
    >
      <div
        class="inline-fields col-6 pl-0"
      >
        <input
          class="form-control"
          type="text"
          value="test1"
        />
      </div>
      <div
        class="icon-buttons col-6"
      >
        <button
          class="icon-button btn btn-secondary"
          type="button"
        >
          <i
            class="bi bi-arrows-expand"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`should collapse and match snapshot 1`] = `
<div>
  <div
    class="expandable-section"
  >
    <div
      class="expandable-section-header"
    >
      <div
        class="inline-fields col-6 pl-0"
      >
        <input
          class="form-control"
          type="text"
          value="test1"
        />
      </div>
      <div
        class="icon-buttons col-6"
      >
        <button
          class="icon-button btn btn-secondary"
          type="button"
        >
          <i
            class="bi bi-arrows-expand"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`should deletable render match snapshot 1`] = `
<div>
  <div
    class="expandable-section"
  >
    <div
      class="expandable-section-header"
    >
      <div
        class="inline-fields col-6 pl-0"
      >
        <input
          class="form-control"
          type="text"
          value="test1"
        />
      </div>
      <div
        class="icon-buttons col-6"
      >
        <button
          class="icon-button btn btn-secondary"
          type="button"
        >
          <i
            class="bi bi-trash"
          />
        </button>
        <button
          class="icon-button btn btn-secondary"
          type="button"
        >
          <i
            class="bi bi-arrows-expand"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`should expand and match snapshot 1`] = `
<div>
  <div
    class="expandable-section"
  >
    <div
      class="expandable-section-header"
    >
      <div
        class="inline-fields col-6 pl-0"
      >
        <input
          class="form-control"
          type="text"
          value="test1"
        />
      </div>
      <div
        class="icon-buttons col-6"
      >
        <button
          class="icon-button btn btn-secondary"
          type="button"
        >
          <i
            class="bi bi-arrows-collapse"
          />
        </button>
      </div>
    </div>
    <div
      class="expandable-section-body"
    >
      <div>
        <input
          class="form-control"
          type="text"
          value="test2"
        />
        <input
          class="form-control"
          type="text"
          value="test3"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`should render match snapshot 1`] = `
<div>
  <div
    class="expandable-section"
  >
    <div
      class="expandable-section-header"
    >
      <div
        class="inline-fields col-6 pl-0"
      >
        <input
          class="form-control"
          type="text"
          value="test1"
        />
      </div>
      <div
        class="icon-buttons col-6"
      >
        <button
          class="icon-button btn btn-secondary"
          type="button"
        >
          <i
            class="bi bi-arrows-expand"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
